polonius-engine/src/pipeline.rs

```diff
@@ -0,0 +1,221 @@
+        }
+
+        O::load_from_db(&facts)
+    }
+
+    /// Check that this pipeline is able to compute the specified outputs if given the specificied
+    /// inputs.
+    ///
+    /// Panics if this requirement is not met.
+    fn validate(&self, inputs: Rels, outputs: Rels) {
+        let mut available: FxHashSet<&str> = Default::default();
+        available.extend(inputs.iter());
+
+        for unit in self.0 {
+            // Ensure that the required inputs have all been computed
+            for input in unit.inputs() {
+                if !available.contains(input) {
+                    panic!(
+                        "`{}` required by {} but not provided by input or preceding computation",
+                        input,
+                        unit.name()
+                    )
+                }
+            }
+            available.extend(unit.outputs());
+        }
+
+        for output in outputs {
+            if !available.contains(output) {
+                panic!(
+                    "Required output `{}` not computed by any computation",
+                    output
+                )
+            }
+        }
+    }
+}
+
+/// An object-safe wrapper around a [`Computation`].
+pub trait ComputationDyn<T: FactTypes> {
+    /// A human-readable name for this computation.
+    fn name(&self) -> &'static str;
+
+    /// The required inputs.
+    fn inputs(&self) -> Rels;
+
+    /// The outputs.
+    fn outputs(&self) -> Rels;
+
+    /// Loads the input for a [`Computation`], runs it, and stores the result.
+    fn compute(&self, db: &mut Db<T>, dump: &mut Dump<'_>);
+}
+
+// `#![feature(associated_type_bounds)]` is required for this impl. Otherwise the type parameters
+// we create to represent for `Input` and `Output` are unbound.
+//
+// FIXME: Can we write this impl without the feature gate?
+impl<C, T> ComputationDyn<T> for C
+where
+    C: for<'db> Computation<T, Input<'db>: LoadFrom<'db, T>, Output: StoreTo<T>>,
+    T: FactTypes,
+{
+    fn name(&self) -> &'static str {
+        readable_typename::<C>()
+    }
+
+    fn inputs(&self) -> Rels {
+        <C::Input<'_>>::RELATIONS
+    }
+
+    fn outputs(&self) -> Rels {
+        <C::Output>::RELATIONS
+    }
+
+    fn compute(&self, db: &mut Db<T>, dump: &mut Dump<'_>) {
+        compute_(self, db, dump)
+    }
+}
+
+/// Loads the input for a [`Computation`], runs it, and stores the result.
+fn compute_<C, T>(computation: &C, db: &mut Db<T>, dump: &mut Dump<'_>)
+where
+    C: for<'db> Computation<T, Input<'db>: LoadFrom<'db, T>, Output: StoreTo<T>>,
+    T: FactTypes,
+{
+    let name = readable_typename::<C>();
+    db.curr_unit = name;
+    info!("Running computation `{}`...", name);
+
+    let input = <C::Input<'_>>::load_from_db(db);
+    dump.unit_start(name);
+    let start_time = Instant::now();
+    let output = computation.compute(input, dump);
+    let end_time = Instant::now();
+    dump.unit_end(name);
+
+    output.store_to_db(db, dump);
```

Comment:
Here's the part that requires GATs/associated type bounds. Moving the lifetime parameter for `Input` to `Computation` itself and changing the signature to the following gives an error:

```rust
fn compute_<'db, C, T>(computation: &C, db: &'db mut Db<T>, dump: &mut Dump<'_>)
where
    C: Computation<'db, T, Input: LoadFrom<'db, T>, Output: StoreTo<T>>,
    T: FactTypes,
```

```
error[E0502]: cannot borrow `*db` as mutable because it is also borrowed as immutable
   --> polonius-engine/src/pipeline.rs:209:5
    |
193 | fn compute_<'db, C, T>(computation: &C, db: &'db mut Db<T>, dump: &mut Dump<'_>)
    |             --- lifetime `'db` defined here
...
202 |     let input = <C::Input>::load_from_db(db);
    |                 ----------------------------
    |                 |                        |
    |                 |                        immutable borrow occurs here
    |                 argument requires that `*db` is borrowed for `'db`
...
209 |     output.store_to_db(db, dump);
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^ mutable borrow occurs here
```

This is expected since data from `db` flows from `input` into `computation` and then `output`. It's perhaps more surprising that this works with a GAT on `Computation::Input`?

Reply:
My understanding is that this works with GATs because neither the implementer of `Computation` nor `Computation::Output` can name the lifetime used in `Computation::Input<'db>`. I'm not very confident in this assessment, however.